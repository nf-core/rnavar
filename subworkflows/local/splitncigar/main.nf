//
// Subworkflow: Run GATK4 SplitNCigarReads with intervals, merge and index BAM file.
//

include { GATK4_SPLITNCIGARREADS } from '../../../modules/nf-core/gatk4/splitncigarreads/main'
include { SAMTOOLS_MERGE         } from '../../../modules/nf-core/samtools/merge/main'
include { SAMTOOLS_INDEX         } from '../../../modules/nf-core/samtools/index/main'

workflow SPLITNCIGAR {
    take:
    bam          // channel: [ val(meta), [ bam ], [bai] ]
    fasta        // channel: [ fasta ]
    fai          // channel: [ fai ]
    dict         // channel: [ dict ]
    intervals    // channel: [ interval_list]

    main:
    def ch_versions       = Channel.empty()

    def bam_interval = bam
        .combine(intervals)
        .map { meta, bam_, bai, intervals_ ->
            [ meta + [interval_count:intervals_ instanceof List ? intervals_.size() : 1], bam_, bai, intervals_ ]
        }
        .transpose(by:3)
        .map { meta, bam_, bai, interval ->
            [ meta + [id:"${meta.id}_${interval.baseName}", sample: meta.id], bam_, bai, interval ]
        }

    GATK4_SPLITNCIGARREADS(bam_interval,
        fasta,
        fai.map{ fai_ -> [[id:'genome'], fai_] },
        dict)

    def bam_splitncigar = GATK4_SPLITNCIGARREADS.out.bam
    ch_versions = ch_versions.mix(GATK4_SPLITNCIGARREADS.out.versions)

    def bam_splitncigar_interval = bam_splitncigar
        .map{ meta, bam_ ->
            def new_meta = meta + [id:meta.sample] - meta.subMap('sample') - meta.subMap('interval_count')
            [ groupKey(new_meta, meta.interval_count), bam_ ]
        }
        .groupTuple()

    SAMTOOLS_MERGE(
        bam_splitncigar_interval,
        fasta,
        fai.map{ fai_ -> [[id:fai_.baseName], fai_] }
    )

    def splitncigar_bam = SAMTOOLS_MERGE.out.bam
    ch_versions = ch_versions.mix(SAMTOOLS_MERGE.out.versions)

    SAMTOOLS_INDEX(splitncigar_bam)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions)

    def splitncigar_bam_indices = SAMTOOLS_INDEX.out.bai
        .mix(SAMTOOLS_INDEX.out.csi)
        .mix(SAMTOOLS_INDEX.out.crai)

    def splitncigar_bam_bai = splitncigar_bam
        .join(splitncigar_bam_indices, failOnDuplicate: true, failOnMismatch: true)

    emit:
    bam_bai     = splitncigar_bam_bai
    versions    = ch_versions
}
