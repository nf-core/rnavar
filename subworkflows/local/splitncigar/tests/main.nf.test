nextflow_workflow {

    name "Test Subworkflow SPLITNCIGAR"
    script "../main.nf"
    workflow "SPLITNCIGAR"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/splitncigar"
    tag "samtools/merge"
    tag "samtools/index"
    tag "gatk4/splitncigarreads"

    test("bam") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.value([
                    [id:"fai"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = Channel.value([
                    [id:"dict"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                ])
                input[4] = Channel.of([[
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.filtered_intervals.interval_list', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.interval_list', checkIfExists: true)
                ]])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bam_bai.collect { [it[0], bam(it[1]).readsMD5, file(it[2].toString()).name] },
                    workflow.out.versions
                ).match() }
            )
        }
    }

    test("cram") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram.crai', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.value([
                    [id:"fai"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = Channel.value([
                    [id:"dict"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                ])
                input[4] = Channel.of([[
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.filtered_intervals.interval_list', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.interval_list', checkIfExists: true)
                ]])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bam_bai.collect { [it[0], bam(it[1]).readsMD5, file(it[2].toString()).name] },
                    workflow.out.versions
                ).match() }
            )
        }
    }

    test("no_intervals") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.value([
                    [id:"fai"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = Channel.value([
                    [id:"dict"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                ])
                input[4] = Channel.of([[]])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bam_bai.collect { [it[0], bam(it[1]).readsMD5, file(it[2].toString()).name] },
                    workflow.out.versions
                ).match() }
            )
        }
    }
}
