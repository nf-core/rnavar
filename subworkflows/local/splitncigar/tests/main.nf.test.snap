{
    "cram": {
        "content": [
            [
                [
                    {
                        "groupSize": 2,
                        "groupTarget": {
                            "id": "test"
                        }
                    },
                    "67daadcc078bb45be360e1f2af46a8e5",
                    "test.bam.bai"
                ]
            ],
            [
                "versions.yml:md5,2ee606c033a70f44ba961106bb4727dd",
                "versions.yml:md5,7809d5e62a9a71923012f7d06dc7a0fb",
                "versions.yml:md5,7809d5e62a9a71923012f7d06dc7a0fb",
                "versions.yml:md5,f2707dd02aa42f4f8154476966707508"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T17:15:55.786758395"
    },
    "no_intervals": {
        "content": [
            [
                [
                    {
                        "groupSize": 0,
                        "groupTarget": {
                            "id": "test"
                        }
                    },
                    "7ebc60b3ccddb8489781f46522bc5da1",
                    "test.bam.bai"
                ]
            ],
            [
                "versions.yml:md5,2ee606c033a70f44ba961106bb4727dd",
                "versions.yml:md5,7809d5e62a9a71923012f7d06dc7a0fb",
                "versions.yml:md5,f2707dd02aa42f4f8154476966707508"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T17:16:38.597763964"
    },
    "bam": {
        "content": [
            [
                [
                    {
                        "groupSize": 2,
                        "groupTarget": {
                            "id": "test"
                        }
                    },
                    "67daadcc078bb45be360e1f2af46a8e5",
                    "test.bam.bai"
                ]
            ],
            [
                "versions.yml:md5,2ee606c033a70f44ba961106bb4727dd",
                "versions.yml:md5,7809d5e62a9a71923012f7d06dc7a0fb",
                "versions.yml:md5,7809d5e62a9a71923012f7d06dc7a0fb",
                "versions.yml:md5,f2707dd02aa42f4f8154476966707508"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T17:15:10.999131603"
    }
}